chunky/src/java/se/llbit/chunky/world/material/TextureMaterial.java

```diff
@@ -16,11 +16,31 @@
  */
 package se.llbit.chunky.world.material;
 
+import org.apache.commons.collections4.map.AbstractReferenceMap;
+import org.apache.commons.collections4.map.ReferenceMap;
 import se.llbit.chunky.resources.Texture;
 import se.llbit.chunky.world.Material;
 
+import java.util.Collections;
+import java.util.Map;
+
 public class TextureMaterial extends Material {
 
+  // Cache the Texture -> TextureMaterial association to prevent creating tons of
+  // TextureMaterial using the same Texture
+  static private final Map<Texture, TextureMaterial> cache = Collections.synchronizedMap(
+    new ReferenceMap<>(AbstractReferenceMap.ReferenceStrength.HARD, AbstractReferenceMap.ReferenceStrength.WEAK)
+  );
+
+  public static TextureMaterial getForTexture(Texture texture) {
+    TextureMaterial mat = cache.get(texture);
+    if(mat == null) {
+      mat = new TextureMaterial(texture);
+      cache.put(texture, mat);
+    }
+    return mat;
+  }
+
   public TextureMaterial(Texture texture) {
```

Comment:
If you want to make sure that the cache is used, you could make this constructor private and check for sign textures by `instance of`` in `getForTexture`.

Good candidates for no caching would probably we skull textures, too.

Reply:
There are some other uses of the constructor that I left untouched, mainly the ones where the `MaterialTexture` was static and always shared without help from the cache.

Will look at skull textures

Reply:
If i'm reading the code correctly, custom skull textures are not supported, only the built-in steve (and monsters) so for now it makes sense to cache it

Reply:
Custom skull textures are supported since 2.3.0

Reply:
Oh yeah sorry, I was reading SkullEntity instead of HeadEntity